File: code/ex347.java
import java.util.*;

class Solution {  
    public List<Integer> topKFrequent(int[] nums, int k) {
        // Build hash table
        Map<Integer, Integer> hash = new HashMap<>();
        for (int num : nums)    hash.put(num, hash.getOrDefault(num, 0) + 1);
        
        // Get results
        List<Integer> res = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : hash.entrySet()) {
            res.add(entry.getKey());
        }
        
        res.sort((a, b) -> hash.get(b) - hash.get(a));       
        
        return res.subList(0, k);
    }
}

File: code/ex347.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> cnt;
        for (int num : nums)    cnt[num]++;
        
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        for (auto& [key, val] : cnt) {
            pq.push({val, key});
            int size = pq.size();
            if (pq.size() > k)    pq.pop();
        }
        
        vector<int> res;
        while (!pq.empty()) {
            res.push_back(pq.top().second);
            pq.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};